File: CN/Experiment 3/CRC.c
#include <stdio.h>
#include <string.h>
char dx[50], gx[10], fx[50], rx[50];
int dxlen, gxlen, fxlen, rxlen;

void xor(char* a, char* b, int len) {
	int i = 1;
	while (i < len) {
		if (a[i] == b[i]) {
			a[i - 1] = '0';
		} else {
			a[i - 1] = '1';
		}
		i++;
	}
	a[i - 1] = '\0';
}

void crc(char* dx, char* gx, char* fx) {
	int i = 0, j = 0;
	while (i < strlen(gx) - 1) {
		dx[strlen(dx) + i] = '0';
		i++;
	}
	dx[strlen(dx) + i] = '\0';
	for (i = 0; i < strlen(gx); i++) {
		fx[i] = dx[i];
	}
	i = 0;
	while (i < strlen(dx) - strlen(gx)) {
		if (fx[0] == '1') {
			xor(fx, gx, strlen(gx));
		} else {
			for (j = 1; j < strlen(gx); j++) {
				fx[j - 1] = fx[j];
			}
		}
		if (i < strlen(dx) - strlen(gx) - 1)
			fx[gxlen - 1] = dx[gxlen + i];
		else
			fx[gxlen - 1] = '\0';
		i++;
	}
	fx[i] = '\0';
}

int main() {
	int i = 0, j = 0;
	printf("Enter the data blocks: ");
	scanf("%s", dx);
	printf("\nEnter the generator polynomial: ");
	scanf("%s", gx);
	dxlen = strlen(dx);
	gxlen = strlen(gx);
	while (i < gxlen - 1) {
		dx[dxlen + i] = '0';
		i++;
	}
	dx[dxlen + i] = '\0';
	dxlen = strlen(dx);
	crc(dx, gx, fx);
	printf("\nCRC code: %s\n", fx);
	i = strlen(dx) - strlen(fx);
	while (i < strlen(dx)) {
		dx[i] = fx[i - strlen(dx) + strlen(fx)];
		i++;
	}
	dx[i] = '\0';
	printf("\nTransmitted data: %s\n", dx);
	printf("\nEnter the received data: ");
	scanf("%s", rx);
	rxlen = strlen(rx);
	crc(rx, gx, fx);
	printf("\nRemainder: %s\n", fx);
	for (i = 0; i < strlen(fx); i++) {
		if (fx[i] == '1') {
			printf("\nError detected\n");
			return 0;
		}
	}
	printf("\nNo error detected\n");
	return 0;
}

File: CN/Experiment 3/CRC.py
# Sender side:
data = input('Enter the data: ')
original_data = data
generator = input('Enter the generator polynomial: ')
data = list(data)
generator = list(generator)
for i in range(len(generator) - 1):
	data.append('0')

def xor(a, b):
	if a == b:
		return '0'
	return '1'

def calcRemainder(data, generator):
	while len(data) >= len(generator):
		if data[0] == '1':
			for i in range(len(generator)):
				data[i] = xor(data[i], generator[i])
		while data and data[0] == '0':
			data.pop(0)
	while len(data) < len(generator) - 1:
		data.insert(0, '0')
	return data

remainder = calcRemainder(data, generator)
crc = original_data + ''.join(remainder)
print('Remainder:', ''.join(remainder))
print('CRC:', crc)
print()

# Receiver side:
received_data = input('Enter the received data: ')
received_generator = input('Enter the generator polynomial: ')
original_dataR = received_data[:len(received_data) - len(received_generator) + 1]
received_data = list(received_data)
received_generator = list(received_generator)
remainderR = calcRemainder(received_data, received_generator)

if '1' in remainderR:
	print('Error detected')
else:
	print('No error detected')
	print('Original data:', original_dataR)